Space charge 2D envelope push

cemitch99 opened the pull request:
This PR implements the linear push of the covariance matrix in the presence of 2D space charge.

- [x] Basic math for space charge push
- [x] Add calculation of beam perveance
- [x] Add beam current user input
- [x] Update space charge flag to allow 2D/3D options
- [x] Adjust location to src/envelope
- [x] Add Python bindings
- [x] Benchmark example
- [x] Update documentation
- [x] Finalize control parameters & update examples

This addresses Issue #826 in the 2D (envelope) case.  Follow-up:  a corresponding model for 3D space charge.

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#839)

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.9.4 → v0.9.6](https://github.com/astral-sh/ruff-pre-commit/compare/v0.9.4...v0.9.6)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

ax3l committed:
Doc: How-To Update (#838)

Move "workflows" to "how-to guides" as in WarpX.
Clean up some wording.

cemitch99 committed:
Initial draft commit.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.H:
```diff
@@ -0,0 +1,45 @@
+#include "particles/ImpactXParticleContainer.H"
+#include "particles/CovarianceMatrix.H"
+
+#include <AMReX_Geometry.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_Vector.H>
+
+#include <unordered_map>
```
```suggestion
#include <AMReX_REAL.H>
```

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,66 @@
+
+#include <AMReX_BLProfiler.H>
+#include <AMReX_REAL.H>       // for Real
+#include <AMReX_SmallMatrix.H>
```
```suggestion
#include <AMReX_SmallMatrix.H>

#include <cmath>

```

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,66 @@
+         *  reduced 2D space charge model, based on the beam covariance matrix.
+         *
+         * @param[in] refpart reference particle
+         * @param[in] cm covariance matrix
```
```suggestion
         * @param[in,out] cm covariance matrix
```

ax3l committed:
Fix `beam.units = static` (#840)

Currently, we only support (and document correctly) "static" units,
but the user was able to also pass "dynamic", upon which we did print
"Dynamic units" but then still used static ones.

This now errors out correctly for anything else than "static" until
we implement another variant.

cemitch99 committed:
Update EnvelopeSpaceChargePush.H

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update EnvelopeSpaceChargePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update EnvelopeSpaceChargePush.cpp

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,68 @@
+
+#include <cmath>
+
+namespace impactx::spacecharge
```
We could potentially move these two files to `src/envelope/spacecharge/...` to avoid confusion with the particle methods, what do you think?

Corresponding C++ namespace then could be:
```suggestion
namespace impactx::envelope::spacecharge
```

cemitch99 replied:
Ok, let's discuss.  There will likely be other reduced space charge models that are not specific to `envelope` tracking, and we need a place to put them also, so we may need to think about a more natural organization.  I propose to leave this here temporarily for the initial implementation and debugging, while we assess.

cemitch99 replied:
I moves the files to `src/envelope`, but I didn't rename the namespace yet, to avoid confusion with the name of the struct `Envelope`.

ax3l replied:
Updated :)

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.H:
```diff
@@ -0,0 +1,41 @@
+     envelope_space_charge2D_push (
+         RefPart const & refpart,
+         Map6x6 & cm,
+         amrex::ParticleReal & current,
+         amrex::ParticleReal & ds
```
Scalar values that are only read are best passed by value in C++:
```suggestion
         amrex::ParticleReal current,
         amrex::ParticleReal ds
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

ax3l commented on src/particles/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,68 @@
+        envelope_space_charge2D_push (
+        RefPart const & refpart,
+        Map6x6 & cm,
+        amrex::ParticleReal & current,
+        amrex::ParticleReal & ds
```
See interface file comment:
```suggestion
        amrex::ParticleReal current,
        amrex::ParticleReal ds
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

ax3l commented on src/tracking/envelope.cpp:
```diff
@@ -77,7 +78,7 @@ namespace impactx
         bool space_charge = false;
         pp_algo.query("space_charge", space_charge);
-        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(!space_charge, "Space charge not yet implemented for envelope tracking.");
+        //AMREX_ALWAYS_ASSERT_WITH_MESSAGE("2D space charge only is implemented for envelope tracking.");
```
Ah, maybe a good time for us to introduce a new parameter for the space charge variants we will implement soon and in the future (2D, 3D, other?).

Something like `algo.space_charge_model = ...` (string?)

cemitch99 replied:
Agreed.

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l committed:
openPMD Beam Input via `Source` Element (#820)

* `Source` Element for openPMD

* Rename Source Files of `BeamMonitor`

* Deduplicate openPMD Logic Helpers

* Docs

* Example

* Doc: Fix Copy-Paste

proy30 committed:
Dashboard: Add tooltips (#843)

* update dashboard __init__.py import
* Add tooltips functionality to non-nested states
* Add tooltips for the custom vselects
* update type hints

ax3l committed:
Update Stub Files

cemitch99 committed:
Remove AMREX_FORCE_INLINE

cemitch99 committed:
Add calculation of beam perveance.

cemitch99 committed:
Fix ParmParse.

cemitch99 committed:
Update EnvelopeSpaceChargePush.cpp

cemitch99 committed:
Update InitDistribution.cpp

cemitch99 committed:
Merge branch 'development' into add_spch2D_envelope

cemitch99 committed:
Input file for benchmark example.

cemitch99 committed:
Correct perveance calculation. Hard-coded current for testing.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/tracking/envelope.cpp:
```diff
@@ -123,6 +124,13 @@ namespace impactx
+                    if (space_charge)
+                    {
+                        // push Covariance Matrix in 2D space charge fields
+                        amrex::ParticleReal current=0.5;  //TODO: This must be set.
```
This is hard-coded to the benchmark example value of 0.5 A, until the ParmParse logic is added.

cemitch99 committed:
Update benchmark example input files.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Fix unused ParmParse variable error.

cemitch99 committed:
Remove unused drift3.

cemitch99 committed:
Add support for current input, some refactoring.

cemitch99 committed:
Make Python current arg optional.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Fix Doxygen in CovarianceMatrix.H

cemitch99 committed:
Pass covariance matrix by reference.

cemitch99 committed:
Fix Doxygen error.

cemitch99 committed:
Another Doxygen error.

cemitch99 committed:
Try to fix Doxygen again.

cemitch99 committed:
Add space_charge_model input.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Move space charge envelope files.

cemitch99 committed:
Add basic documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -42,6 +42,9 @@ Initial Beam Distributions
   bunch charge
 
+* ``beam.current`` (``float``, in A)
+  beam current, used only if space_charge_model = "2D"
```
Just adding some verbatim highlights:
```suggestion
  beam current, used only if ``algo.space_charge_model = "2D"``
```

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -824,6 +827,20 @@ Multigrid-specific numerical options:
+
+  * ``3D``: Space charge forces are computed in three dimensions, assuming the beam is bunched.
+
+    Currently, this model is supported only in particle mode (when algo.track = "particles").
```
```suggestion
    Currently, this model is supported only in envelope mode (when ``algo.track = "envelope"``).

  * ``3D``: Space charge forces are computed in three dimensions, assuming the beam is bunched.

    Currently, this model is supported only in particle mode (when ``algo.track = "particles"``).
```

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -180,13 +180,14 @@ Collective Effects & Overall Simulation Parameters
       :param int npart: number of particles to draw
 
-   .. py:method:: init_envelope(ref, distr)
+   .. py:method:: init_envelope(ref, distr, intensity)
```
Do we want to make this
```suggestion
   .. py:method:: init_envelope(ref, distr, intensity=None)
```
so users do not need to provide an `intensity` if we do not use space charge?

ax3l replied:
Updated :)

ax3l replied:
This also fixed the failing test for existing (non-space-charge) envelope tests.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -180,13 +180,14 @@ Collective Effects & Overall Simulation Parameters
 
       :param ref: the reference particle (object from :py:class:`impactx.RefPart`)
       :param distr: distribution function (object from :py:mod:`impactx.distribution`)
+      :param float intensity: the beam intensity, given as bunch charge (C) or beam current (A)
```
Depends on 3D/2D solve if it is a charge or current?

ax3l replied:
updated doc string

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -1269,3 +1269,19 @@ add_impactx_test(examples-scraping.py
+    examples/fodo_space_charge/analysis_fodo_envelope_sc.py
+    OFF  # no plot script yet
+)
+add_impactx_test(fodo-envelope-sc.py
```
I think that is the naming scheme we used for the FODO tests:
```suggestion
add_impactx_test(FODO.envelope.sc
    examples/fodo_space_charge/input_fodo_envelope_sc.in
    ON   # ImpactX MPI-parallel
    examples/fodo_space_charge/analysis_fodo_envelope_sc.py
    OFF  # no plot script yet
)
add_impactx_test(FODO.envelope.sc.py
```

ax3l commented on examples/fodo_space_charge/README.rst:
```diff
@@ -0,0 +1,75 @@
+.. _examples-fodo-envelope-sc:
+
+FODO Cell with 2D Space Charge using Envelope Tracking
+=======================================================
```
```suggestion
FODO Cell with 2D Space Charge using Envelope Tracking
======================================================
```

ax3l commented on examples/fodo_space_charge/README.rst:
```diff
@@ -0,0 +1,75 @@
+with 2D space charge included.  The parameters are those described in:
+
+R.D. Ryne et al, "A Test Suite of Space-Charge Problems for Code Benchmarking,"
+in Proc. EPAC 2004, Lucerne, Switzerland:  KV Beam in a FODO Channel
```
```suggestion
R.D. Ryne et al, `"A Test Suite of Space-Charge Problems for Code Benchmarking," <https://accelconf.web.cern.ch/e04/PAPERS/WEPLT047.PDF>`__
in Proc. EPAC 2004, Lucerne, Switzerland:  KV Beam in a FODO Channel
```

ax3l commented on examples/fodo_space_charge/README.rst:
```diff
@@ -0,0 +1,75 @@
+
+   .. literalinclude:: plot_fodo.py
+      :language: python3
+      :caption: You can copy this file from ``examples/fodo/plot_fodo.py``.
```
All file names and directories in this section still need to be updated to the new `fodo_space_charge/...` path and name.

ax3l replied:
updated by Chad

ax3l commented:
Can you update the first table in `docs/source/usage/how_to_run.rst` for the last column?
![image](https://github.com/user-attachments/assets/f6861200-96ce-4d5a-854a-4eeb6d8cd20c)

cemitch99 committed:
Update how_to_run table.

ax3l commented on src/envelope/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+target_sources(lib
+  PRIVATE
+)
+
```
```suggestion
```

ax3l commented on src/envelope/spacecharge/EnvelopeSpaceChargePush.H:
```diff
@@ -0,0 +1,41 @@
+#include <AMReX_REAL.H>
+
+
+namespace impactx::spacecharge
```
You could move the `envelope_` prefix to a namespace here
```suggestion
namespace impactx::envelope::spacecharge
```
and remove the prefix from the function name below.

That way we mirror the directory structure, too, which is good.

(Same change then needs to go in the `.cpp` file and the locations we call this function from.

ax3l replied:
updated.

ax3l commented on src/envelope/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace impactx::spacecharge
+{
+        void
+        envelope_space_charge2D_push (
+        [[maybe_unused]] RefPart const & refpart,
+        Map6x6 & cm,
+        amrex::ParticleReal current,
+        amrex::ParticleReal ds
```
```suggestion
    void
    envelope_space_charge2D_push (
        RefPart const & refpart,
        Map6x6 & cm,
        amrex::ParticleReal current,
        amrex::ParticleReal ds
```

ax3l replied:
updated

ax3l commented on src/envelope/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,67 @@
+        Map6x6 R = Map6x6::Identity();
+
+        // physical constants and reference quantities
+        amrex::ParticleReal const c = ablastr::constant::SI::c;
+        amrex::ParticleReal const ep0 = ablastr::constant::SI::ep0;
+        amrex::ParticleReal const pi = ablastr::constant::math::pi;
```
I think you can also write:
```suggestion
        using ablastr::constant::SI::c;
        using ablastr::constant::SI::ep0;
        using ablastr::constant::math::pi;
```
:pie:

ax3l replied:
done :)

ax3l commented on src/envelope/spacecharge/EnvelopeSpaceChargePush.cpp:
```diff
@@ -0,0 +1,67 @@
+        amrex::ParticleReal const betgam3 = std::pow(betgam,3);
+
+        // evaluate the beam space charge perveance from current
+        amrex::ParticleReal const IA = 4.0_prt*pi*ep0*mass*pow(c,3)/charge;
```
C -> C++ function (for overloads that support both `double` and `float` for `c`)
```suggestion
        amrex::ParticleReal const IA = 4.0_prt*pi*ep0*mass*std::pow(c,3)/charge;
```

ax3l replied:
updated

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -40,8 +40,9 @@ namespace impactx::initialization
-    create_covariance_matrix (
+    Envelope
+    create_envelope (
+        amrex::ParticleReal const current,
```
If we want to make current/charge optional, we could make this the 2nd argument and use `std::optional`, which is similar to Python's `=None`.

ax3l replied:
There is a small inconsistency with naming current/charge/intensity here that we can unify.

ax3l replied:
updated

ax3l reviewed (commented):
This looks awesome!

Added small inline comments.

ax3l committed:
Verbatim

ax3l committed:
CMake: Rename Examples

ax3l committed:
CMake: Remove Empty Command

ax3l committed:
Readme Formatting & Link

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -471,9 +476,22 @@ namespace impactx
+            pp_algo.query("space_charge_model", space_charge_model);
+            amrex::ParticleReal intensity = 0.0; // bunch charge (C) for 3D model, beam current (A) for 2D model
+
+            if (space_charge_model == "3D") {
+                pp_dist.query("charge", intensity);
+                throw std::runtime_error("3D space charge model not yet implemented in envelope mode.");
+            } else if (space_charge_model == "2D") {
+                pp_dist.query("current", intensity);
+            } else {
+                throw std::runtime_error("Unknown space_charge_model (use '2D' or '3D') ");
+            }
```
I think I want to unify this with `algo.space_charge`. There is no need for two parameters, we can just do `algo.space_charge = false` (default), `= 2D` and `= 3D`.

cemitch99 replied:
This makes sense, although it will break the existing space charge examples, where `algo.space_charge = true`.  In that case, maybe we should push the change through as a follow-up, since it also impacts the logic outside the envelope tracking?

ax3l replied:
I think I can implement it in a backwards compatible way, will give it a try tomorrow.

ax3l replied:
I pushed a backwards compatible change and it currently passes CI. This is what existing users will see (True defaults now to 3D space charge and raises a warning to use the new syntax.)

I will push another commit now that updates all examples to the new syntax, so people see and use that going forward. After a few months (year), we can drop support for the old syntax. I added an inline TODO where we will remove it.

ax3l committed:
Apply Changes from Review, Fix Tests

cemitch99 committed:
Update benchmark README.

ax3l committed:
Backwards Compatible `algo.space_charge` evolution

ax3l committed:
Python: `Envelope` Class

cemitch99 commented on examples/aperture/input_aperture.in:
```diff
@@ -40,7 +40,7 @@ collimator.aperture_y = 1.5e-3
 # Algorithms
 ###############################################################################
 algo.particle_shape = 2
-algo.space_charge = false
```
I have one concern about this recent change.  With `algo.space_charge = false`, it is obvious to the user that space charge is not active.  Without this line, I think it is too easy to forget that space charge is off by default, _especially_ when a nonzero bunch charge is provided.  This is especially important in the examples.  If we keep three string input options "2D", "3D", and "false", then we could keep this line.

ax3l replied:
I was wondering about that, good point.

Off was and is the default; I agree we can keep it explicit to avoid any confusion.

cemitch99 commented on src/python/ImpactX.cpp:
```diff
@@ -227,14 +230,32 @@ void init_ImpactX (py::module& m)
+                        pp_algo.add("space_charge", std::string("3D"));
+                    } else {
+                        py::print("sim.space_charge = False is deprecated, please use space_charge = \"off\" (or skip for default)");
+                        pp_algo.add("space_charge", std::string("off"));
```
Yes, I see that the "off" string is supported, so I would keep this in the example input files.

ax3l replied:
Agreed. And we can still use `false` as well.

The problem why I need to internally call this `off` in the enum is that `false` is a C++ keyword.

ax3l replied:
Ah, can name it `False` internally (uppercase)

ax3l committed:
Update all Examples

ax3l commented on src/tracking/particles.cpp:
```diff
@@ -68,13 +69,16 @@ namespace impactx
+        auto space_charge = get_space_charge_algo();
         if (verbose > 0) {
-            amrex::Print() << " Space Charge effects: " << space_charge << "\n";
+            amrex::Print() << " Space Charge effects: " << amrex::getEnumNameString(space_charge) << "\n";
```
I think I want to strip the `True_` prefix with a helper function, because then we exactly mirror the input we receive from the user when we print to the terminal.

ax3l replied:
Done.

ax3l commented on src/tracking/envelope.cpp:
```diff
@@ -78,21 +79,23 @@ namespace impactx
         if (verbose > 0)
         {
-            amrex::Print() << " Space Charge effects: " << space_charge << "\n";
+            amrex::Print() << " Space Charge effects: " << amrex::getEnumNameString(space_charge) << "\n";
```
Strip `True_` prefix

ax3l replied:
Done.

ax3l commented on src/tracking/envelope.cpp:
```diff
@@ -46,12 +48,14 @@ namespace impactx
-        auto & cm = amr_data->track_envelope.m_cm.value();
+        auto & env = amr_data->track_envelope.m_env.value();
+        auto & cm = env.m_env;
+        auto & current = env.m_beam_intensity;
```
Double check if we skip the space charge on zero beam intensity.

ax3l replied:
Also: raise a warning if a user explicitly turns on space charge (or CSR) and then sets the beam intensity to zero.

ax3l replied:
see below.

ax3l replied:
Warning added.

ax3l commented on src/tracking/particles.cpp:
```diff
@@ -68,13 +69,16 @@ namespace impactx
-        amrex::ParmParse const pp_algo("algo");
-        bool space_charge = false;
-        pp_algo.query("space_charge", space_charge);
+        auto space_charge = get_space_charge_algo();
```
Double check if we skip the space charge on zero beam charge.

ax3l replied:
Also: raise a warning if a user explicitly turns on space charge (or CSR) and then sets the charge to zero.

ax3l replied:
Ah perfect, we have this warning already:
```
**** WARNINGS ******************************************************************
* GLOBAL warning list  after  [ FIRST STEP ]
*
* --> [!  ] [ElectrostaticSolver] [raised once]
*     Max norm of rho is 0
*     @ Raised by: ALL
*
* --> [!  ] [ImpactX::add_particles] [raised once]
*     The bunch charge is set to zero. ImpactX will run with zero-weighted
*     particles. Did you mean to set the space charge algorithm to off instead?
*     @ Raised by: ALL
*
********************************************************************************
```

ax3l committed:
Improve Status Print

And add doxygen and improve case matching.

ax3l committed:
Envelope: Warn on Zero Current (2D SC)
